pallets/did/src/lib.rs

```diff
@@ -1512,6 +1527,17 @@ pub mod pallet {
 			// `take` calls `kill` internally
 			let did_entry = Did::<T>::take(&did_subject).ok_or(Error::<T>::NotFound)?;
 
+			// Make sure this check happens after the line where we check if a DID exists,
+			// else we would start getting `CannotDelete` errors when we should be getting
+			// `NotFound`.
+			ensure!(
+				<<T::DidLifecycleHooks as DidLifecycleHooks<T>>::DeletionHook as DidDeletionHook<T>>::can_delete(
+					&did_subject,
+				)
+				.is_ok(),
+				Error::<T>::CannotDelete
+			);
```

Comment:
Not sure if this is feasible in substrate, but I'd much rather raise errors in the hook and have them propagated, so that we can inform callers on which deletion conditions have not been met.

Reply:
I can see this approach hitting a wall though if we can't come up with a way to define errors that are not already defined in any of the pallets. Substrate seems to be somewhat limited in that regard.

Reply:
The only thing we can do is return `u8` or `u16` error codes, which are returned by the runtime hook. Then we would need to come up with a scheme that would be able to encode in 8 or 16 bits all the different "components" that are linked to the DID, to be able to return an aggregate error of all the pending stuff instead of returning the error of the first item that was encountered.

Reply:
Since we will add runtime APIs to ease deletion, I would say that a dry run of this extrinsic + a call to the API to return the list of call is sufficient to solve this issue. I would not over-complicate the runtime code.